### PR TITLE
bugfix: Escape java home path on windows

### PR DESCRIPTION
**Pull request by tgodzik**
This was reported by a student on EPFL. JSON doesn't allow for `\`

@tanishiking let's merge this before the release

Fixes https://github.com/scalameta/metals/issues/3968

### CODE REVIEW

**Comment by tanishiking** [comment]
Looks good, but I'll test it locally at least it works on macOS just in case.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala]: [note]
If `ensureDesiredVersion` reconnect to the bloop server, we don't need to `ensureDesiredJvmSettings` because it's already restarted and jvm settings should be already desired state.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:232]: I decided that we don't need to ask in case the json file is managed by Bloop, since that will confuse new users. Especially that we set the javaHome at the start.

File: metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala:1318]: Let's not show both messages at the same time, anyway usually it should be just one thing.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: Ach, missed it!

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: And should be fixed.

**Review by tanishiking — APPROVED** [review]
I left one nitpick, otherwise it's LGTM!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (dos65) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:158]: There is a way to simlify duplication inside pattern matching
```scala
      val fields: List[(String, ujson.Value)] =
        List(
          maybeJavaHome.map(v => "javaHome" -> ujson.Str(v)),
          maybeBloopJvmProperties.map(opts =>
            "javaOptions" -> ujson.Arr(opts.map(ujson.Str(_)))
          )
        ).flatten
      val obj = ujson.Obj.from(fields)
      val propertiesString = ujson.write(obj)
```

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (dos65) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:159]: Is it ok to leave it empty?
it looks like like previously the empty was written as an object `{}`

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:158]: I was thinking about it, but it seemed a bit clearer this way with everything written explicitely.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:159]: This can't really happen since we don't write anything if both are None.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (zmerr) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:214]: why is this removed? This was used for telling the user why bloop was to be restarted, especially when it was caused by changing Metals java home.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:218]: It seems to me that when the restart is done due to update to Metals Java home, the users need a way to know this is going to affect their Bloop java home too. Don’t they?

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:158]: Ok sure, not a big deal.

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: I’m afraid the format ujson might use for producing the json might not be exactly compatible with the form in Bloop docs. I remember that’s manual serialization was used in the code in the first place for producing the json. It’s good to double check the produced result and check also if it can be properly consumed both during reading and by Bloop.

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: Changed!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:391]: I decided that since we are changing things in the settings it just adds additional work to user to then click to restart Bloop server. Especially, that a lot of users might not understand it fully.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:214]: Same as above.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: We can't anyway use manual serialization (as that leads to a lot of issues) and since this a valid json it should all work correctly. I double checked it. Also, Bloop actually uses a json library for reading, so this should be ok.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:218]: The main problem is that we will also do this at the start and we would always show it to the user even if it's a new user that doesn't know anything about the Java home etc. So I decide that it's better to hide it currently instead. Otherwise I expect a lot of questions from confused users.

**Review by dos65 — APPROVED** [review]

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: Please, let me pull the branch and double check if it can find the expected value.

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: Sorry Tomasz, in local testing of the branch, Metals cannot detect that the value written in the Bloop json 
```
{"javaOptions": [-Xmx1G]}
```
is the same as the one assigned to Bloop Jvm Properties as `-Xmx1G` and shows the popup!
To reproduce, first manually change the file to the content shown above. then go to settings and set bloop jvm properties to `Xmx1G`
The expected behavior is to do nothing. Instead a popup dialogue is shown.

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: Oh, it is written automatically with quotation marks, but when I put the quotes there:

```
{"javaOptions": ["-Xmx1G", "Xms100M”]}
```
it still shows the dialogue even if the requested value is equal to the content of the file.
Not sure why.

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: I think it is unrelated to json. Now I am trying to find the root of the issue, since this was supposed to be solved in the previous PR.

**Comment by tanishiking** [comment]
I'll wait the [release](https://github.com/scalameta/metals/pull/3957) until this is merged.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: Ok then we will merge this one and if you could explain the other problem in an issue or create a PR that would be awesome!

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:143]: I just added the trim to make sure it all works even when whitespaces are added

File: metals/src/main/scala/scala/meta/internal/metals/Messages.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/Messages.scala:355]: I reverted back to asking the user because otherwise when multiple metals servers were running this would result in race conditions

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:378]: This is to make sure that we don't restart Bloop after setting bloop.json for the first time.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:215]: Changed the order here to first ask then update to avoid race conditions which might be what @zmerr was seeing

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala]: I think I found the issue, mentioned it in the other comment.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (zmerr) [inline metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:306]: so `maybeBloopGlobalJsonJvmProperties` is always defined?

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:306]: Yes, if it's an empty list it results in no options set for Bloop

Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:306]: So it covers also the scenario when we want to remove the properties, right?

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (zmerr) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:380]: this should only happen when the javaHome that Bloop was using was already the same as that of Metals, or as the one we set for Metals in the setting.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:260]: IT might be a None if it doesn't exist in the json file, but we treat it the same as an empty list.

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:306]: Yes, that works

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:380]: This will always happen at the start, when no bloop.json was set up previously.

Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:380]: So now, if

- there's no `$HOME/.bloop/bloop.json` and
- Whether there's `metals.bloopJvmProperties` defined or not

Metals will always write the global bloop setting (when connect to bloop launcher) file to `$HOME/.bloop/bloop.json` where `javaHome` will be inferred java home location by `locate-java-home`.

---

I'm a bit concerned about the following scenario:
- Metals write the global bloop file to `$HOME/.bloop/bloop.json` where `javaHome` is javaHome of currently using java (inferred by `locate-java-home`)
- User upgrade java version (and now `locate-java-home` locate the new java's javahome.
- Metals won't update the global bloop setting file, because the user setting of Metals didn't change (and `workspace/didChangeConfiguration` doesn't fire).
  - Or the user declines Metals to write bloop global file.
- While Metals uses new java, bloop still has old javaHome in the global bloop setting file, and bloop will continue using old Java.

It might be better to
- (1) every time Metals server starts, write JVM properties to bloop global json file even if there exists a file (but I'm not sure how should we handle the existing file :thinking:)
- (2) don't write java home if a user didn't specify java home explicitly via user settings.

**Review by tanishiking — CHANGES_REQUESTED** [review]
I commented about one concern

File: metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala
Comment (tanishiking) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:380]: @tgodzik 
Oh does workspace/didChangeConfigration will  be fired (at least by VSCode) on server starts, even if the setting has changed before Metals server starts?
In that case, it's all good :+1:

 
> The only issue will be on other editors if Metals was started and the setting was later updated manually

I'm not sure, but other clients also should fire. Anyway, it would be better to implement (1) as you mentioned, if it's handy. If it's complexed, we can go as it is, at least for now

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:380]: > Oh does workspace/didChangeConfigration will be fired (at least by VSCode) on server starts, even if the setting has changed before Metals server starts?
> In that case, it's all good

Yeah, it should be sent correctly.

> I'm not sure, but other clients also should fire. Anyway, it would be better to implement (1) as you mentioned, if it's handy. If it's complexed, we can go as it is, at least for now

So another thing is that if set javaHome before starting Bloop, we would have to restart it in case it's running, which is a bit more complex and would be outside the user's influence. For now I would leave at is and maybe improve the behaviour later on?

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala:380]: Coming to think of it you currently need to connect to Bloop to restart it, so we would need to invoke Bloop CLI for that (which could be a part of a different issue though) https://github.com/scalameta/metals/issues/3542

**Review by tanishiking — APPROVED** [review]

**Comment by tanishiking** [comment]
Well, I realized that when I manually set `javaHome` of the workspace setting (of VSCode), Metals also write the workspace javaHome setting to `$HOME/.bloop/bloop.json`.
However, workspace settings should be limited to the workspace, and it shouldn't change the global settings I guess. Is this an expected behavior?

**Comment by tanishiking** [comment]
Anyway @tgodzik great work :tada:

**Pull request merged**
